test/e2e/storage/csimock/csi_honor_pv_reclaim_policy.go

```diff
@@ -0,0 +1,192 @@
+	"k8s.io/kubernetes/test/e2e/feature"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epv "k8s.io/kubernetes/test/e2e/framework/pv"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+	admissionapi "k8s.io/pod-security-admission/api"
+	"k8s.io/utils/ptr"
+)
+
+var _ = utils.SIGDescribe("CSI Mock honor pv reclaim policy", feature.HonorPVReclaimPolicy, framework.WithFeatureGate(features.HonorPVReclaimPolicy), func() {
+	f := framework.NewDefaultFramework("csi-mock-honor-pv-reclaim-policy")
+	f.NamespacePodSecurityLevel = admissionapi.LevelPrivileged
+	m := newMockDriverSetup(f)
+
+	ginkgo.Context("CSI honor pv reclaim policy using mock driver", func() {
+		ginkgo.It("Dynamic provisioning should honor pv delete reclaim policy", func(ctx context.Context) {
+			m.init(ctx, testParameters{
+				registerDriver:             true,
+				enableHonorPVReclaimPolicy: true,
+				reclaimPolicy:              ptr.To(v1.PersistentVolumeReclaimDelete),
+			})
+			ginkgo.DeferCleanup(m.cleanup)
+
+			_, pvc := m.createPVC(ctx)
+
+			ginkgo.By(fmt.Sprintf("Waiting for PVC %s to be bound", pvc.Name))
+			pvs, err := e2epv.WaitForPVClaimBoundPhase(ctx, f.ClientSet, []*v1.PersistentVolumeClaim{pvc}, framework.ClaimProvisionTimeout)
+			framework.ExpectNoError(err, "failed to wait for PVC to be bound")
+			gomega.Expect(pvs).To(gomega.HaveLen(1), "expected 1 PV to be bound to PVC, got %d", len(pvs))
+
+			pv := pvs[0]
+			ginkgo.By(fmt.Sprintf("PVC %s is bound to PV %s", pvc.Name, pv.Name))
+			gomega.Expect(pv.Spec.PersistentVolumeReclaimPolicy).To(gomega.Equal(v1.PersistentVolumeReclaimDelete),
+				"expected PV %s to have reclaim policy %s, got %s", pv.Name, v1.PersistentVolumeReclaimDelete, pv.Spec.PersistentVolumeReclaimPolicy)
+			// For dynamic provisioning, the PV should be created with the deletion protection finalizer.
+			gomega.Expect(pv.Finalizers).To(gomega.ContainElement(storagehelpers.PVDeletionProtectionFinalizer),
+				"expected PV %s to have finalizer %s", pv.Name, storagehelpers.PVDeletionProtectionFinalizer)
+
+			ginkgo.By(fmt.Sprintf("Deleting PV %s", pv.Name))
+			err = f.ClientSet.CoreV1().PersistentVolumes().Delete(ctx, pv.Name, metav1.DeleteOptions{})
+			framework.ExpectNoError(err, "failed to delete PV %s", pv.Name)
+
+			ginkgo.By(fmt.Sprintf("Deleting PVC %s", pvc.Name))
+			err = f.ClientSet.CoreV1().PersistentVolumeClaims(pvc.Namespace).Delete(ctx, pvc.Name, metav1.DeleteOptions{})
+			framework.ExpectNoError(err, "failed to delete PVC %s", pvc.Name)
+
+			ginkgo.By(fmt.Sprintf("Waiting for PV %s to be deleted", pv.Name))
+			err = e2epv.WaitForPersistentVolumeDeleted(ctx, f.ClientSet, pv.Name, framework.Poll, 2*time.Minute)
+			framework.ExpectNoError(err, "failed to wait for PV to be deleted")
+
+			ginkgo.By(fmt.Sprintf("Verifying that the driver received DeleteVolume call for PV %s", pv.Name))
+			gomega.Expect(m.driver.GetCalls(ctx)).To(gomega.ContainElement(gomega.HaveField("Method", gomega.Equal("DeleteVolume"))))
```

Comment:
In the below test i see `err = e2epv.WaitForPersistentVolumeDeleted(ctx, f.ClientSet, pv.Name, framework.Poll, 2*time.Minute)`, i am wondering if we can make the same changes here

Reply:
sorry, I can not understand your idea and why we need this change.

Reply:
The test only checks if DeleteVolume was called, by calling `e2epv.WaitForPersistentVolumeDeleted` you can be sure the whole PV is deleted as well.

Reply:
@deepakkinni `e2epv.WaitForPersistentVolumeDeleted` was called in L76 before checking if DeleteVolume was called.